packages/backend/test/username.test.ts

```diff
@@ -0,0 +1,123 @@
+import test from 'ava'
+import { startServer } from './environment'
+import fetch from 'node-fetch'
+import { hashOne } from '@unirep/crypto'
+import { ethers } from 'ethers'
+
+import {
+    signIn,
+    signUp,
+    setUsername,
+    epochTransition,
+    userStateTransition,
+    setSameUsername,
+    waitForBackendBlock,
+    genUsernameProof,
+} from './utils'
+
+const EPOCH_LENGTH = 20000
+
+test.before(async (t: any) => {
+    const context = await startServer({ epochLength: EPOCH_LENGTH / 1000 })
+    Object.assign(t.context, context)
+})
+
+test('should set a username', async (t: any) => {
+    // sign up and sign in user
+    const { iden, commitment } = await signUp(t)
+    await signIn(t, commitment)
+
+    // first set a username
+    // pre-image by default is 0
+    await setUsername(t, iden, 0, 'initial-test-username123')
+
+    await new Promise((r) => setTimeout(r, EPOCH_LENGTH))
+
+    // execute the epoch transition
+    const prevEpoch = await t.context.unirep.currentEpoch()
+    await epochTransition(t)
+    for (;;) {
+        await new Promise((r) => setTimeout(r, 1000))
+        const findEpoch = await t.context.db.findOne('Epoch', {
+            where: { number: Number(prevEpoch) },
+        })
+        if (findEpoch) break
+    }
+
+    // user state transition
+    await userStateTransition(t, iden)
+
+    // change the username to something else
+    await setUsername(
+        t,
+        iden,
+        'initial-test-username123',
+        'second-test-username123'
+    )
+
+    t.pass()
+})
+
+test.serial(
+    'should fail to set the username that is already taken',
+    async (t: any) => {
+        // sign up and sign in user
+        const { iden, commitment } = await signUp(t)
+        await signIn(t, commitment)
+
+        // first set a username
+        // pre-image by default is 0
+        await setUsername(t, iden, 0, 'username123')
+
+        await new Promise((r) => setTimeout(r, EPOCH_LENGTH))
+
+        // execute the epoch transition
+        const prevEpoch = await t.context.unirep.currentEpoch()
+        await epochTransition(t)
+        for (;;) {
+            await new Promise((r) => setTimeout(r, 1000))
+            const findEpoch = await t.context.db.findOne('Epoch', {
+                where: { number: Number(prevEpoch) },
+            })
+            if (findEpoch) break
+        }
+
+        // user state transition
+        await userStateTransition(t, iden)
+
+        // try to change the username to the same one
+        await setSameUsername(t, iden, 'username123', 'username123')
```

Comment:
There's no need to use a custom function here. We are expecting the backend function to throw an error so you can write it like this:

```js
try {
  await setUsername(t, iden, 'username123', 'username123')
  t.fail() // we expect the function to throw, if it doesn't fail the test
} catch (err) {
  t.true(err.toString().startsWith('/post error'))
}
```

Reply:
@vimwitch regarding the comment, do you mean "we expect the function to throw an error, if it *does* fail the test"?

Reply:
No, I meant when we’re testing a failure case the `setUsername` function _should_ throw an error because the contract function reverts which causes the backend to return a 4xx response. So we want to make sure it throws because if it doesn’t we have a problem.